pkg/sdk/cadence/cron.go

```diff
@@ -0,0 +1,195 @@
+// limitations under the License.
+
+package cadence
+
+import (
+	"context"
+	"time"
+
+	"emperror.dev/errors"
+	"go.uber.org/cadence/.gen/go/shared"
+	"go.uber.org/cadence/client"
+)
+
+// CronConfiguration encapsulates information about a cron workflow.
+type CronConfiguration struct {
+	CadenceClient                client.Client
+	CronInstanceType             CronInstanceType
+	CronSchedule                 string
+	ExecutionStartToCloseTimeout time.Duration
+	TaskListName                 string
+	Workflow                     string
+	WorkflowArguments            []interface{}
+}
+
+// NewCronConfiguration instantiates a cron configuration from the specified
+// values.
+func NewCronConfiguration(
+	cadenceClient client.Client,
+	cronInstanceType CronInstanceType,
+	cronSchedule string,
+	executionStartToCloseTimeout time.Duration,
+	taskListName string,
+	workflow string,
+	workflowArguments ...interface{},
+) CronConfiguration {
+	return CronConfiguration{
+		CadenceClient:                cadenceClient,
+		CronInstanceType:             cronInstanceType,
+		CronSchedule:                 cronSchedule,
+		ExecutionStartToCloseTimeout: executionStartToCloseTimeout,
+		TaskListName:                 taskListName,
+		WorkflowArguments:            workflowArguments,
+		Workflow:                     workflow,
+	}
+}
+
+// CronWorkflowID returns the ID of the cron workflow based on its configuration.
+func (cronConfig CronConfiguration) CronWorkflowID() (cronWorkflowID string) {
+	switch cronConfig.CronInstanceType {
+	case CronInstanceTypeDomain:
+		return cronConfig.TaskListName + "-cron-" + cronConfig.Workflow
+	}
+
+	return string(cronConfig.CronInstanceType) + "-cron-" + cronConfig.Workflow
+}
+
+// StartCronWorkflow initiates a Cadence cron workflow from the specified type.
+//
+// WARNING: the current implementation allows only single instance crons (one
+// per Cadence server).
+func (cronConfig CronConfiguration) StartCronWorkflow(ctx context.Context) (err error) {
+	state, err := cronConfig.WorkflowState(ctx)
+	if err != nil {
+		return errors.Wrap(err, "querying workflow state failed")
+	}
+
+	switch state {
+	case CronWorkflowStateScheduled: // Note: nothing to do.
+		return nil
+	case CronWorkflowStateScheduledOutdated: // Note: restart required.
+		err = cronConfig.CadenceClient.TerminateWorkflow(
+			ctx,
+			cronConfig.CronWorkflowID(),
+			"",
+			"cron workflow schedule requires an update",
+			nil,
+		)
+		if err != nil {
+			return errors.Wrap(err, "terminating cron workflow failed")
+		}
+	}
+
+	cronWorkflowOptions := client.StartWorkflowOptions{
+		ID:                           cronConfig.CronWorkflowID(),
+		TaskList:                     cronConfig.TaskListName,
+		ExecutionStartToCloseTimeout: cronConfig.ExecutionStartToCloseTimeout,
+		CronSchedule:                 cronConfig.CronSchedule,
+		Memo: map[string]interface{}{ // Note: CronSchedule is not directly retrievable (version 0.13.4-0.15.0).
```

Comment:
How hard would it be to migrate away from this once we upgrade to a version that supports retrieving the cron schedule?

The question is: should we hold off until we upgrade?

Reply:
I think it should be between easy and trivial.

Once the cron schedule becomes retrievable, only the generic `CronConfiguration.WorkflowState()` part needs to be adjusted to use the "official retrieval" for obtaining the original cron schedule to decide whether the schedule changed.
These implementation details are behind the generic cron configuration abstraction layer and currently used for determining the requirement for (re)scheduling, nothing would change from the perspective of the cron workflow user/initiator code.

In theory the retrieval implementation should reside next to the generic Memo collection so hopefully this only means throwing out the Memo handling code and referencing a different field from the same workflow execution description object.

Worst case we need to call a different function of the Cadence client, if the cron schedule only appears there and not in the currently used `Client.DescribeWorkflowExecution()` result object.

However there is no indication of this feature being implemented anywhere in the near future.